scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/TypescriptCodeGenerationService.java

```diff
@@ -0,0 +1,140 @@
+import org.jetbrains.annotations.NotNull;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+public class TypescriptCodeGenerationService {
+  public record ActivityType(String name, Map<String, ValueSchema> parameters) {}
+  public record ResourceType(String name, String type, ValueSchema schema) {}
+  public record MissionModelTypes(Collection<ActivityType> activityTypes, Collection<ResourceType> resourceTypes) {}
+
+  private final MerlinService merlinService;
+
+  public TypescriptCodeGenerationService(MerlinService merlinService) {
+    this.merlinService = merlinService;
+  }
+
+  public String generateTypescriptTypesForPlan(final PlanId planId) {
+    try {
+      return generateTypescriptTypesFromMissionModel(this.merlinService.getMissionModelTypes(planId));
+    } catch (IOException e) {
+      throw new Error("Could not fetch mission model types", e);
+    }
+  }
+  public static String generateTypescriptTypesFromMissionModel(final MissionModelTypes missionModelTypes) {
+    final var activityTypeCodes = new ArrayList<ActivityTypeCode>();
+    for (final var activityType : missionModelTypes.activityTypes()) {
+      activityTypeCodes.add(getActivityTypeInformation(activityType));
+    }
+    var result = "/** Start Codegen */\n";
+    for (final var activityTypeCode : activityTypeCodes) {
+      result += activityTypeCode.declaration() + "\n";
+    }
+    result += "export const ActivityTemplates = {\n";
+    for (final var activityTypeCode : activityTypeCodes) {
+      result += activityTypeCode.implementation();
+    }
+    result += "}\n/** End Codegen */";
+    return result;
+  }
+
+  private record ActivityTypeCode(String declaration, String implementation) {}
+
+  private static ActivityTypeCode getActivityTypeInformation(final ActivityType activityType) {
+    return new ActivityTypeCode(
+        String.format("interface %s extends ActivityTemplate {}\n", activityType.name()),
+        String.format("""
+          %s: function %s(
+            name: string,
+            args: {
+            %s
+            }): %s {
+            return {
+              name,
+              activityType: '%s',
+              args: args,
+            };
+          },
+        """, activityType.name(), activityType.name(), generateActivityParameterTypes(activityType).indent(2), activityType.name(), activityType.name())
+    );
+  }
+
+  @NotNull
+  private static String generateActivityParameterTypes(final ActivityType activityType) {
+    final var result = new ArrayList<>();
+    for (final var param : activityType.parameters().entrySet().stream().sorted(Map.Entry.comparingByKey()).toList()) {
+      final var name = param.getKey();
+      final var valueSchema = param.getValue();
+      result.add(String.format("%s: %s,", name, valueSchemaToTypescriptType(valueSchema)));
+    }
+    return String.join("\n", result.toArray(new String[0]));
```

Comment:
No need to change this but this could also be written as:
```java
result.stream().collect(Collectors.joining(System.getProperty("line.separator")))
```
Which is a longer one-liner 😂 ... so maybe nevermind.